File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l <= r:
            m = (l + r) // 2
            if nums[m] == target:
                return m
            if nums[m] >= nums[l]:
                # Left side is sorted
                if target >= nums[l] and target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            else:
                # Right side is sorted
                if target > nums[m] and target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1
        return -1
        

File: Data Structures & Algorithms/eating-bananas/submission-2.cpp
// Binary Search on the answer.
// Time: O(n * log m), where n = piles.length() and m = max(piles)
// Space: O(1)
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        // Maximum possible eating speed.
        int maxPile = 0;
        for (int pile : piles) {
            maxPile = max(maxPile, pile);
        }

        int left = 1;
        int right = maxPile;

        // Stores the best valid speed found so far.
        int k = maxPile;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            // Calculate hours needed if eating at speed = mid.
            long long totalHours = 0;
            for (int pile : piles) {
                // ceil(pile / mid)
                totalHours += (1LL * pile + mid - 1) / mid;
            }

            if (totalHours <= h) {
                // mid works, try to find a smaller valid speed.
                k = mid;
                right = mid - 1;
            } else {
                // mid is too slow, increase the speed.
                left = mid + 1;
            }
        }

        return k;
    }
};